File: src/pytra/std/os_path.py
"""pytra.std.os_path: extern-marked os.path subset with Python runtime fallback."""


from pytra.std import extern

import os.path as __path


@extern
def join(a: str, b: str) -> str:
    return __path.join(a, b)


@extern
def dirname(p: str) -> str:
    return __path.dirname(p)


@extern
def basename(p: str) -> str:
    return __path.basename(p)


@extern
def splitext(p: str) -> tuple[str, str]:
    return __path.splitext(p)


@extern
def abspath(p: str) -> str:
    return __path.abspath(p)


@extern
def exists(p: str) -> bool:
    return __path.exists(p)

File: src/runtime/cpp/std/os_path.cpp
#include "std/os_path.h"
#include <filesystem>

str join(const str& a, const str& b) {
    if (a.empty()) return b;
    if (b.empty()) return a;
    const char tail = a.std().back();
    if (tail == '/' || tail == '\\') return a + b;
    return a + "/" + b;
}

str dirname(const str& p) {
    const ::std::string s = p.std();
    const ::std::size_t pos = s.find_last_of("/\\");
    if (pos == ::std::string::npos) return str("");
    return str(s.substr(0, pos));
}

str basename(const str& p) {
    const ::std::string s = p.std();
    const ::std::size_t pos = s.find_last_of("/\\");
    if (pos == ::std::string::npos) return str(s);
    return str(s.substr(pos + 1));
}

::std::tuple<str, str> splitext(const str& p) {
    const ::std::string s = p.std();
    const ::std::size_t sep = s.find_last_of("/\\");
    const ::std::size_t dot = s.find_last_of('.');
    if (dot == ::std::string::npos) return ::std::tuple<str, str>{str(s), str("")};
    if (sep != ::std::string::npos && dot < sep + 1) return ::std::tuple<str, str>{str(s), str("")};
    return ::std::tuple<str, str>{str(s.substr(0, dot)), str(s.substr(dot))};
}

str abspath(const str& p) {
    return str(::std::filesystem::absolute(::std::filesystem::path(p.std())).generic_string());
}

bool exists(const str& p) {
    return ::std::filesystem::exists(::std::filesystem::path(p.std()));
}
